### PR TITLE
Core Protocol Compatibility

### PR DESCRIPTION
**Pull request by narayanprusty**
## Description

- Integrated Space ID of Binance oracle to fetch feed registry address
- Fetch real price of VAI
- Added fork tests for vBNB, vXVS, VAI and vLTC
- Integrated ACM

## Checklist

- [x] I have updated the documentation to account for the changes in the code.
- [x] If I added new functionality, I added tests covering it.
- [x] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [x] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/oracles/BoundValidator.sol [hunk 1]
```diff
@@ -18,14 +18,18 @@ struct ValidateConfig {
...
     /// @notice validation configs by asset
     mapping(address => ValidateConfig) public validateConfigs;
 
     /// @notice vBNB address
     /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
     address public immutable vBnb;
 
+    /// @notice VAI address
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    address public immutable vai;
```
Comment (chechu) [inline contracts/oracles/BoundValidator.sol:31]: We are not using the `vai` variable in this contract. Should we take it into account in `validatePriceWithAnchorPrice`?

File: contracts/oracles/TwapOracle.sol [hunk 2]
```diff
@@ -161,8 +182,16 @@ contract TwapOracle is OwnableUpgradeable, TwapInterface {
...
-        address asset = address(vToken) == vBnb ? WBNB : VBep20Interface(vToken).underlying();
+        address asset;
+
+        if (address(vToken) == vBnb) {
+            asset = WBNB;
+        } else if (address(vToken) == vai) {
+            asset = vai;
+        } else {
+            asset = VBep20Interface(vToken).underlying();
+        }
```
Comment (chechu) [inline contracts/oracles/TwapOracle.sol:193]: We could define an internal function to encapsulate this if/else statement

File: contracts/Governance/AccessControlManager.sol [hunk 3]
```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
[the remark this answers is not in the document]
Comment (0xlucian) [reply contracts/Governance/AccessControlManager.sol:1]: We can import only the interface and reuse the ACM deployed, indeed.

File: contracts/ResilientOracle.sol [hunk 4]
```diff
@@ -269,7 +292,16 @@ contract ResilientOracle is OwnableUpgradeable, PausableUpgradeable, ResilientOr
...
-        address asset = address(vToken) == vBnb ? BNB_ADDR : VBep20Interface(vToken).underlying();
+        address asset;
+
+        if (address(vToken) == vBnb) {
+            asset = BNB_ADDR;
+        } else if (address(vToken) == vai) {
+            asset = vai;
+        } else {
+            asset = VBep20Interface(vToken).underlying();
+        }
```
[the remark this answers is not in the document]
Comment (0xlucian) [reply contracts/ResilientOracle.sol:303]: +1

File: contracts/oracles/BinanceOracle.sol [hunk 5]
```diff
@@ -3,31 +3,54 @@ pragma solidity 0.8.13;
...
+    }
+
+    /**
+     * @notice Uses Space ID to fetch the feed registry address
+     * @return feedRegistryAddress Address of binance oracle feed registry.
      */
-    function initialize(FeedRegistryInterface feed) public initializer {
-        feedRegistry = feed;
+    function getFeedRegistryAddress() public view returns (address) {
+        bytes32 nodeHash = 0x94fe3821e0768eb35012484db4df61890f9a6ca5bfa984ef8ff717e73139faff;
```
Comment (0xlucian) [inline contracts/oracles/BinanceOracle.sol:47]: is this `nodeHash` same for all networks?

File: deploy/1-deploy-oracles.ts [hunk 6]
```diff
@@ -26,18 +28,28 @@ const func: DeployFunction = async function ({ getNamedAccounts, deployments, ne
   const networkName = network.name === "bscmainnet" ? "bscmainnet" : "bsctestnet";
 
   const vBNBAddress = ADDRESSES[networkName].vBNBAddress;
+  const VAIAddress = ADDRESSES[networkName].VAIAddress;
   const WBNBAddress = ADDRESSES[networkName].WBNBAddress;
 
+  await deploy("AccessControlManager", {
```
Comment (0xlucian) [inline deploy/1-deploy-oracles.ts:34]: We want to reuse the already deployed ACM, so wouldn't it be more suitable if we either import the deployment from IL repo, or just hardcode the ACM address for each network?

File: contracts/Governance/AccessControlManager.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (narayanprusty) [reply contracts/Governance/AccessControlManager.sol:1]: Discussed with @chechu about this. We will deploy the ACM in Oracle as we need to get it audited. Once ACM is audited as part of Oracle repo we will copy the contract to IL repo.

File: deploy/1-deploy-oracles.ts [hunk 7]
```diff
@@ -8,14 +8,16 @@ const ADDRESSES = {
   bsctestnet: {
     vBNBAddress: testnetDeployments.Contracts.vBNB,
     WBNBAddress: "0xae13d989daC2f0dEbFf460aC112a837C89BAa7cd",
+    VAIAddress: "0x05889b75Ae9Aa53AaE49562432201a0c93BaCA0b",
```
Comment (chechu) [inline deploy/1-deploy-oracles.ts:11]: I think `0x05889b75Ae9Aa53AaE49562432201a0c93BaCA0b` is a very old VAI token in testnet. We can use the included in the core package ([0x5fFbE5302BadED40941A403228E6AD03f93752d9](https://testnet.bscscan.com/address/0x5fFbE5302BadED40941A403228E6AD03f93752d9)), that seems newer.

```suggestion
    VAIAddress: testnetDeployments.Contracts.VAI,
```

File: deploy/1-deploy-oracles.ts [hunk 8]
```diff
@@ -8,14 +8,16 @@ const ADDRESSES = {
...
     pythOracleAddress: "0xd7308b14BF4008e7C7196eC35610B1427C5702EA",
-    binanceFeedRegistryAddress: "0x999DD49FeFdC043fDAC4FE12Bb1e4bb31cB4c47B",
+    sidRegistryAddress: "0xfFB52185b56603e0fd71De9de4F6f902f05EEA23",
   },
   bscmainnet: {
     vBNBAddress: mainnetDeployments.Contracts.vBNB,
     WBNBAddress: "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c",
-    pythOracleAddress: "",
-    binanceFeedRegistryAddress: "",
+    VAIAddress: "0x4BD17003473389A42DAF6a0a729f6Fdb328BbBd7",
```
Comment (chechu) [inline deploy/1-deploy-oracles.ts:18]:
```suggestion
    VAIAddress: mainnetDeployments.Contracts.VAI,
```

File: test/fork/core-compatibility.ts [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    resillientOracle.address,
+    "setTokenConfig(TokenConfig)",
+    deployer.address,
+  );
+
+  return { resillientOracle, boundValidator, chainlinkOracle, twapOracle, pythOracle, binanceOracle };
+}
+
+describe("Core protocol", async () => {
+  let resillientOracle: ResilientOracle;
```
Comment (chechu) [inline test/fork/core-compatibility.ts:109]: Typo. `resillientOracle` vs `resilientOracle`

File: contracts/oracles/BinanceOracle.sol (hunk 5 above)
Comment (narayanprusty) [reply contracts/oracles/BinanceOracle.sol:47]: Yes. You can find out more here: https://oracle.binance.com/docs/knowledge-base/space-id/

File: deploy/1-deploy-oracles.ts (hunk 8 above)
Comment (narayanprusty) [reply deploy/1-deploy-oracles.ts]: Fixed

File: deploy/1-deploy-oracles.ts (hunk 7 above)
Comment (narayanprusty) [reply deploy/1-deploy-oracles.ts]: Fixed

File: contracts/oracles/TwapOracle.sol (hunk 2 above)
Comment (narayanprusty) [reply contracts/oracles/TwapOracle.sol]: Fixed

File: contracts/oracles/BoundValidator.sol (hunk 1 above)
Comment (narayanprusty) [reply contracts/oracles/BoundValidator.sol:31]: Fixed

File: contracts/ResilientOracle.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (narayanprusty) [reply contracts/ResilientOracle.sol]: Fixed

**Review by chechu — APPROVED** [review]

**Pull request merged**